Modules/posixmodule.c

```diff
@@ -13535,46 +13539,33 @@ struct constdef {
 };
 
 static int
-conv_confname(PyObject *arg, int *valuep, struct constdef *table,
-              size_t tablesize)
-{
-    if (PyLong_Check(arg)) {
-        int value = PyLong_AsInt(arg);
-        if (value == -1 && PyErr_Occurred())
-            return 0;
-        *valuep = value;
-        return 1;
-    }
-    else {
-        /* look up the value in the table using a binary search */
-        size_t lo = 0;
-        size_t mid;
-        size_t hi = tablesize;
-        int cmp;
-        const char *confname;
-        if (!PyUnicode_Check(arg)) {
-            PyErr_SetString(PyExc_TypeError,
-                "configuration names must be strings or integers");
-            return 0;
-        }
-        confname = PyUnicode_AsUTF8(arg);
-        if (confname == NULL)
-            return 0;
-        while (lo < hi) {
-            mid = (lo + hi) / 2;
-            cmp = strcmp(confname, table[mid].name);
-            if (cmp < 0)
-                hi = mid;
-            else if (cmp > 0)
-                lo = mid + 1;
-            else {
-                *valuep = table[mid].value;
-                return 1;
+conv_confname(PyObject *module, PyObject *arg, int *valuep, const char *tablename)
+{
+    if (PyUnicode_Check(arg)) {
+        PyObject *table = PyObject_GetAttrString(module, tablename);
+        if (table != NULL) {
+            arg = PyObject_GetItem(table, arg);
+            if (arg == NULL) {
+                PyErr_SetString(
```

Comment:
Call `PyErr_Clear();` first to clear the existing exception.

Reply:
The documentation could be clearer about this, but it looks like the `PyErr_Set...` functions all go through `PyErr_SetRaisedException`, which clears the existing exception automatically.

Reply:
It presumably creates a chained exception though, right? I think we don't want that here.

Reply:
It doesn't create a chained exception:
```
Python 3.14.0a1+ (heads/simplify-conv-confname-dirty:2a3e50a5587, Nov  9 2024, 23:42:24) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.confstr("whatever")
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    os.confstr("whatever")
    ~~~~~~~~~~^^^^^^^^^^^^
ValueError: unrecognized configuration name
>>> 
```

Like it says in the [C API documentation](https://docs.python.org/3/c-api/exceptions.html):

> The error indicator is not the result of [sys.exc_info()](https://docs.python.org/3/library/sys.html#sys.exc_info). The former corresponds to an exception that is not yet caught (and is therefore still propagating), while the latter returns an exception after it is caught (and has therefore stopped propagating).

An exception can only chain after it's been caught, but in this case the KeyError raised by `PyObject_GetItem` is never caught. `PyErr_SetString` will decref and replace it.